packages/dashboard-message-bus/lib/DashboardMessageBus.ts

```diff
@@ -0,0 +1,178 @@
+import WebSocket from "isomorphic-ws";
+import delay from "delay";
+import { EventEmitter } from "events";
+import {
+  base64ToJson,
+  broadcastAndAwaitFirst,
+  broadcastAndDisregard,
+  createMessage,
+  jsonToBase64
+} from "./utils";
+import { promisify } from "util";
+
+interface UnfulfilledRequest {
+  socket: WebSocket;
+  data: WebSocket.Data;
+}
+
+export class DashboardMessageBus extends EventEmitter {
+  private requestsServer: WebSocket.Server;
+  private listenServer: WebSocket.Server;
+  private clientSockets: WebSocket[] = [];
+  private listeningSockets: WebSocket[] = [];
+
+  private unfulfilledRequests: Map<string, UnfulfilledRequest> = new Map([]);
+
+  constructor(
+    public requestsPort: number,
+    public listenPort: number,
+    public host: string = "0.0.0.0"
+  ) {
+    super();
+  }
+
+  start() {
+    this.listenServer = new WebSocket.Server({
+      host: this.host,
+      port: this.listenPort
+    });
+
+    this.listenServer.on("connection", (newListener: WebSocket) => {
+      newListener.on("close", () => {
+        this.logToClients("Listener disconnected", "connections");
+
+        this.listeningSockets = this.listeningSockets.filter(
+          listener => listener !== newListener
+        );
+        this.terminateIfNoConnections();
+      });
+
+      // Process all backlogged (unfulfilled) requests on new listener connection.
+      this.unfulfilledRequests.forEach(({ socket, data }) =>
+        this.processRequest(socket, data, [newListener])
+      );
+
+      this.logToClients("Listener connected", "connections");
+
+      this.listeningSockets.push(newListener);
+    });
+
+    this.requestsServer = new WebSocket.Server({
+      host: this.host,
+      port: this.requestsPort
+    });
+
+    this.requestsServer.on("connection", (newClient: WebSocket) => {
+      newClient.on("message", (data: WebSocket.Data) => {
+        this.processRequest(newClient, data, this.listeningSockets);
+      });
+
+      newClient.on("close", () => {
+        this.logToClients("Client disconnected", "connections");
+
+        this.clientSockets = this.clientSockets.filter(
+          client => client !== newClient
+        );
+        this.clearClientRequests(newClient);
+        this.terminateIfNoConnections();
+      });
+
+      this.logToClients("Client connected", "connections");
+
+      this.clientSockets.push(newClient);
+    });
+  }
+
+  async ready() {
+    if (this.listeningSockets.length > 0) return;
+    await delay(1000);
+    await this.ready();
+  }
+
+  async terminate() {
+    await promisify(this.requestsServer.close.bind(this.requestsServer))();
+    await promisify(this.listenServer.close.bind(this.listenServer))();
+    this.emit("terminate");
+  }
+
+  private async processRequest(
+    socket: WebSocket,
+    data: WebSocket.Data,
+    listeners: WebSocket[]
+  ) {
+    if (typeof data !== "string") return;
```

Comment:
❓ I'm assuming we expect Data to be some json object?  What case does this guard against?

Reply:
All messages (log, invalidate, dashboard-provider) are JSON.stringified and then base64 encoded before they're sent over the wire, and then they're decoded at the other side of the wire. 

So we expect all of the messages to come in as a `"string"` type. But it's technically possible to send `Buffer` data through the websocket connection. Essentially this line just *ignores* that buffer data.

Reply:
Why not just convert buffer to string?

Reply:
We can do that as well. Assume it's utf8?

Reply:
yeah, that seems reasonable if it's not too much effort.

Reply:
Sure, will do!